src/core/manager.c

```diff
@@ -3211,6 +3212,80 @@ static void manager_serialize_gid_refs(Manager *m, FILE *f) {
+        int copy;
+
+        /* We don't actually need the instructions or other data, since this is only used on the other side
+         * for BPF limbo, which just requires the program type, cgroup path, and kernel-facing BPF file
+         * descriptor. We don't even need to know what unit or directive it's attached to, since we're just
+         * going to expire it after coldplug. */
+
+        assert(f);
+        assert(p);
+
+        /* The program should be set up by now, otherwise it shouldn't be present at all */
+        assert(p->kernel_fd > 0);
+        assert(p->attached_path);
+
+        copy = fdset_put_dup(fds, p->kernel_fd);
+        if (copy < 0)
+                return log_error_errno(copy, "Failed to add file descriptor to serialization set: %m");
+
+        return serialize_item_format(f, "bpf-limbo", "%i %i %i \"%s\"",
+                                     copy, p->prog_type, p->attached_type, cescape(p->attached_path));
+}
+
+static void deserialize_limbo_bpf_program(Manager *m, FDSet *fds, const char *value) {
+        _cleanup_free_ char *raw_fd = NULL, *raw_pt = NULL, *raw_at = NULL, *raw_cgpath = NULL;
+        BPFProgram *p;
+        int fd, r, prog_type, attached_type;
+
+        assert(m);
+        assert(value);
+
+        r = extract_first_word(&value, &raw_fd, NULL, 0);
+        if (r <= 0 || safe_atoi(raw_fd, &fd) < 0 || fd < 0 || !fdset_contains(fds, fd)) {
+                log_error("Failed to parse bpf-limbo FD: %s", value);
+                return;
+        }
+
+        r = extract_first_word(&value, &raw_pt, NULL, 0);
+        if (r <= 0 || safe_atoi(raw_pt, &prog_type) < 0) {
+                log_error("Failed to parse bpf-limbo program type: %s", value);
+                return;
+        }
+
+        r = extract_first_word(&value, &raw_at, NULL, 0);
+        if (r <= 0 || safe_atoi(raw_at, &attached_type) < 0) {
+                log_error("Failed to parse bpf-limbo attached type: %s", value);
+                return;
+        }
+
+        r = extract_first_word(&value, &raw_cgpath, NULL, EXTRACT_CUNESCAPE | EXTRACT_UNQUOTE);
+        if (r <= 0) {
+                log_error_errno(r, "Failed to parse attached path for BPF limbo FD %s: %m", value);
+                return;
+        }
+
+        r = bpf_program_new(prog_type, &p);
+        if (r < 0) {
+                log_error_errno(r, "Failed to create BPF limbo program");
+                return;
+        }
+
+        /* Just enough to free it when the time is right, this does not have enough information be used as a
+         * real BPFProgram. */
+        p->attached_type = attached_type;
+        p->kernel_fd = fdset_remove(fds, fd);
+        p->attached_path = strdup(raw_cgpath);
+
+        r = set_ensure_put(&m->bpf_limbo_progs, NULL, p);
+        if (r < 0) {
+                log_error_errno(r, "Failed to register BPF limbo program for FD %s", value);
+                (void) bpf_program_unref(p);
```

Comment:
Not a matter, but I prefer to use `_cleanup_(bpf_program_unrefp) BPFProgram *p = NULL;`, and `TAKE_PTR(p);` on success.

Reply:
Unless you have any objections, I'll keep it this way for now, just because I find it easier to reason about the object lifetimes this way. :-)